### PR TITLE
Change quantizer to account for input tensor's memory format.

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42840 Enable input pointer caching in XNNPACK integration.
* **#42178 Change quantizer to account for input tensor's memory format.**
* #42008 Call qnnpack's conv setup only if input pointer has changed.
* #42007 Refactor qconv to reduce allocations.
* #42006 Simple caching allocator for CPU.

Summary:
This otherwise introduces unnecessary calls to contiguous in the rest of
the network, where certain ops want channels last format.

Test Plan:
Quantization tests.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22796479](https://our.internmc.facebook.com/intern/diff/D22796479)

### CODE REVIEW

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 1]
```diff
@@ -104,18 +104,18 @@ Tensor PerTensorAffineQuantizer::quantize(Tensor rtensor) {
   // quantizer that can be reused, so I'm using intrusive_from_this here
   Tensor qtensor = new_qtensor(
       rtensor.sizes(),
-      rtensor.options().dtype(scalar_type_),
+      rtensor.options().dtype(scalar_type_).memory_format(rtensor.suggest_memory_format()),
       intrusive_from_this());
 
-  rtensor = rtensor.contiguous();
+  rtensor = rtensor.contiguous(rtensor.suggest_memory_format());
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp:110]: is this going to be a no-op?

Comment (jerryzh168) [reply aten/src/ATen/quantized/Quantizer.cpp:110]: or this is still making the tensor contiguous?

File: aten/src/ATen/quantized/Quantizer.cpp [hunk 2]
```diff
@@ -104,18 +104,18 @@ Tensor PerTensorAffineQuantizer::quantize(Tensor rtensor) {
   // quantizer that can be reused, so I'm using intrusive_from_this here
   Tensor qtensor = new_qtensor(
       rtensor.sizes(),
-      rtensor.options().dtype(scalar_type_),
+      rtensor.options().dtype(scalar_type_).memory_format(rtensor.suggest_memory_format()),
```
Comment (jerryzh168) [inline aten/src/ATen/quantized/Quantizer.cpp:107]: do we also need to make qtensor contiguous then? or it is created as a contiguous tensor

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/quantized/Quantizer.cpp:110]: Ideally, it should become no-op.

File: aten/src/ATen/quantized/Quantizer.cpp (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/quantized/Quantizer.cpp:107]: I think by default it does that. I remember new_qtensor does refresh_contiguous, so I am assuming it makes it contiguous in the memory format it is allocated in.

**Comment by jerryzh168** [comment]
looks like there are some failing tests

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 3]
```diff
@@ -2493,8 +2493,6 @@ void quantize_tensor_per_tensor_affine_cpu(
 #if defined(__ARM_NEON__) || defined(__aarch64__)
   AT_DISPATCH_QINT_TYPES(
       qtensor.scalar_type(), "quantize_tensor_per_tensor_affine_cpu", [&]() {
-        TORCH_CHECK(
-            rtensor.is_contiguous(), "Float tensor should be contiguous");
         const float* const rdata = rtensor.data_ptr<float>();
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2496]: I don't understand this.  It's usually not safe to call `data_ptr` on a non-contiguous tensor.  Maybe the check should be that rtensor is contiguous in the memory format of qtensor?  And what if qtensor is not contiguous?

**Review by dreiss — CHANGES_REQUESTED** [review]
I'm not sure the existing test are sufficient to test this.  At the very least, I think we need tests for quantizing channels-first, channels-last, and fully non-contiguous input tensors.  Unless those are already there and I'm just missing them.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 3 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2496]: Oh yes. Thanks for the catch. Dont know what was I thinking removing this, maybe it will come back.

**Comment by kimishpatel** [comment]
> I'm not sure the existing test are sufficient to test this. At the very least, I think we need tests for quantizing channels-first, channels-last, and fully non-contiguous input tensors. Unless those are already there and I'm just missing them.

That sounds good. Let's check with Jerry. @jerryzh168, do you know if quantization unit tests covers all these cases. If not I will add.

**Comment by jerryzh168** [comment]
> > I'm not sure the existing test are sufficient to test this. At the very least, I think we need tests for quantizing channels-first, channels-last, and fully non-contiguous input tensors. Unless those are already there and I'm just missing them.
> 
> That sounds good. Let's check with Jerry. @jerryzh168, do you know if quantization unit tests covers all these cases. If not I will add.

right we don't have tests for different memory formats, we always make Tensor contiguous before.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 4]
```diff
@@ -2572,17 +2624,39 @@ void dequantize_tensor_per_channel_affine_cpu(
         int64_t channel = rtensor.size(axis);
         auto scales_data = scales.data_ptr<double>();
         auto zero_points_data = zero_points.data_ptr<int64_t>();
+        TORCH_CHECK(
+            qtensor.is_contiguous(qtensor.suggest_memory_format()),
+            "Quantized tensor should be contiguous");
+        TORCH_CHECK(
+            rtensor.is_contiguous(qtensor.suggest_memory_format()),
+            "Float tensor should be contiguous "
+            "in same memory format as quantizd tensor");
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2633]: can we create a helper function for these two checks?

File: test/quantization/test_quantized_tensor.py [hunk 5]
```diff
@@ -232,6 +241,53 @@ def quantize_c(data, scales, zero_points):
         self.assertTrue(np.allclose(qr.int_repr(), quantize_c(r, scales, zero_points)))
         self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
 
+        # Check 4D tensor with 2 different memory formats.
+        r = torch.rand(3, 2, 4, 5, dtype=torch.float) * 4 - 2
+        scales = torch.tensor([0.2, 0.03], dtype=torch.double)
+        zero_points = torch.tensor([5, 10], dtype=torch.long)
+        axis = 1
+
+        def quantize_c_4d(data, scales, zero_points):
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:250]: can you make this function general to all dimensions? we can probably reshape the tensor to 1d, do computation in 1d tensor and then reshape it back i think.

File: test/quantization/test_quantized_tensor.py [hunk 6]
```diff
@@ -232,6 +241,53 @@ def quantize_c(data, scales, zero_points):
...
+                        for m in range(5):
+                            res[i][j][l][m] = \
+                                np.clip(np.round(data[i][j][l][m] / scales[j]) + zero_points[j], quant_min, quant_max)
+            return res
+        for memory_format in [torch.contiguous_format, torch.channels_last]:
+            r = r.contiguous(memory_format=memory_format)
+            qr = torch.quantize_per_channel(r, scales, zero_points, axis, torch.quint8)
+            rqr = qr.dequantize()
+            self.assertTrue(np.allclose(qr.int_repr(), quantize_c_4d(r, scales, zero_points)))
+            self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:265]: and merge the test code as well

File: torch/testing/_internal/common_quantization.py [hunk 7]
```diff
@@ -961,7 +961,7 @@ def forward(self, x):
         out = self.conv2(out)
         out = torch.nn.functional.max_pool2d(out, 2, 2)
         out = self.cat.cat([out, out])
-        out = out.view(-1, 3 * 2 * 2)
+        out = out.reshape(-1, 3 * 2 * 2)
```
Comment (jerryzh168) [inline torch/testing/_internal/common_quantization.py:964]: why is this change needed?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 8]
```diff
@@ -2500,7 +2519,12 @@ void quantize_tensor_per_tensor_affine_cpu(
   AT_DISPATCH_QINT_TYPES(
       qtensor.scalar_type(), "quantize_tensor_per_tensor_affine_cpu", [&]() {
         TORCH_CHECK(
-            rtensor.is_contiguous(), "Float tensor should be contiguous");
+            rtensor.is_contiguous(rtensor.suggest_memory_format()),
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2522]: how do we guarantee that these functions are always passed contiguous tensors as inputs?

Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2522]: It is already broken if we are not passing contiguous tensors, because kernels extract raw pointer and operate on them. This check is to ensure that the kernels get what they expect.

File: torch/testing/_internal/common_quantization.py (hunk 7 above)
Comment (kimishpatel) [reply torch/testing/_internal/common_quantization.py:964]: Because `view` is broken for channels_last format and there is not easy way to fix that. Although not clear why this PR exposes that particularly.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 4 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:2624]: Just for checks? I know there are a few places we are doing checks, but it seems strange to have a separate two-line function to just do asserts.

File: test/quantization/test_quantized_tensor.py (hunk 5 above)
Comment (kimishpatel) [reply test/quantization/test_quantized_tensor.py:250]: We can't do 1D tensor because then we have to deconstruct channel dimension to apply per channel quant. But I think we can combine, h, w and d dimensions and then use just one function.

File: test/quantization/test_quantized_tensor.py [hunk 9]
```diff
@@ -257,6 +307,41 @@ def quantize_ref(data, scales, zero_points):
         self.assertTrue(np.allclose(qr.int_repr(), ref))
         self.assertTrue(np.allclose(r.numpy(), dequant_tensor.numpy(), atol=1))
 
+        def _quantize_per_channel_ref_nd(data, scales, zero_points):
+            dims = data.size()
+            data = data.view(-1, dims[1], np.prod(dims[2:]))
+            res = torch.empty_like(data)
+            quant_min, quant_max = 0, 255
+            for i in range(res.size()[0]):
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:315]: nit: res.size(0)

File: test/quantization/test_quantized_tensor.py [hunk 10]
```diff
@@ -257,6 +307,41 @@ def quantize_ref(data, scales, zero_points):
         self.assertTrue(np.allclose(qr.int_repr(), ref))
         self.assertTrue(np.allclose(r.numpy(), dequant_tensor.numpy(), atol=1))
 
+        def _quantize_per_channel_ref_nd(data, scales, zero_points):
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:310]: this function looks very similar to the one in test_qtensor_quantize_per_channel, can we merge them

File: test/quantization/test_quantized_tensor.py [hunk 11]
```diff
@@ -257,6 +307,41 @@ def quantize_ref(data, scales, zero_points):
...
+        # Check 5D tensor.
+        r = torch.rand(3, 2, 4, 5, 7, dtype=torch.float) * 4 - 2
+
+        for memory_format in [torch.contiguous_format, torch.channels_last_3d]:
+            ref_res = _quantize_per_channel_ref_nd(r, scales, zero_points)
+            r = r.contiguous(memory_format=memory_format)
+            qr = torch.quantize_per_channel(r, scales, zero_points, axis, torch.quint8)
+            rqr = qr.dequantize()
+            self.assertTrue(np.allclose(qr.int_repr(), ref_res))
+            self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:344]: checks looks pretty much the same as well, can we merge them into a local function?

File: test/quantization/test_quantized_tensor.py [hunk 12]
```diff
@@ -235,6 +244,47 @@ def quantize_c(data, scales, zero_points):
...
+        zero_points = torch.tensor([5, 10], dtype=torch.long)
+        axis = 1
+
+        for memory_format in [torch.contiguous_format, torch.channels_last_3d]:
+            ref_res = _quantize_per_channel_ref_nd(r, scales, zero_points)
+            r = r.contiguous(memory_format=memory_format)
+            qr = torch.quantize_per_channel(r, scales, zero_points, axis, torch.quint8)
+            rqr = qr.dequantize()
+            self.assertTrue(np.allclose(qr.int_repr(), ref_res))
+            self.assertTrue(np.allclose(r.numpy(), rqr.numpy(), atol=2 / np.min(scales.numpy())))
```
Comment (jerryzh168) [inline test/quantization/test_quantized_tensor.py:286]: same here

File: test/quantization/test_quantized_tensor.py (hunk 10 above)
Comment (kimishpatel) [reply test/quantization/test_quantized_tensor.py:310]: I did that. But there was a reason why I reverted that. Now it is escaping me. I will try to merge, then it will come back.

Comment (kimishpatel) [reply test/quantization/test_quantized_tensor.py:310]: OK, so now I remember. In one of them rounding happens before while the other rounding happens after adding zero point. So could probably add a parameter to the function to chose one over the other, but I opted for just keeping two copies.

Comment (kimishpatel) [reply test/quantization/test_quantized_tensor.py]: Ok done.

**Review by jerryzh168 — APPROVED** [review]
Thanks, had a few inline nit comments
